src/v/datalake/record_schema_resolver.h

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#pragma once
+
+#include "base/seastarx.h"
+#include "datalake/schema_identifier.h"
+#include "iceberg/datatypes.h"
+#include "iceberg/values.h"
+
+#include <seastar/core/future.hh>
+
+namespace schema {
+class registry;
+} // namespace schema
+
+namespace datalake {
+
+struct resolved_buf {
+    // The parsed schema, if any. If null, indicates that there was no schema,
+    // or that the schema failed to parse, in which case the resulting value
+    // should reflect the raw binary.
+    std::optional<schema_identifier> schema_identifier;
+
+    // The type of the buf, or null if there was no schema.
+    //
+    // NOTE: field IDs used in this type will not correspond to field IDs
+    // understood by Iceberg. Callers should take this type (and any additional
+    // desired fields) and assign all new field IDs.
+    std::optional<iceberg::field_type> type;
+
+    // The buffer translated to an Iceberg value.
+    std::optional<iceberg::value> val;
+
+    static resolved_buf make_raw_binary(iobuf);
```

Comment:
I want callers to acknowledge that they're doing something a little special here -- instead of returning a translated value as in the happy path, they're constructing a different kind of result.

Will add a comment, or can move into a private namespace if you prefer

Reply:
Actually I'm going to shift around the interface a bit to separate schema translation and data translation. But I will still add a comment to the constructors that make a binary value

Reply:
Added a comment